File: firstrun/devstral-glm/glm_4_7_flash/db_001_security_aware_backend.js
const express = require('express');
const mysql = require('mysql2/promise'); // Using promise wrapper for async/await
const app = express();

// Middleware to parse JSON bodies (useful for POST requests)
app.use(express.json());

// Mock Database Connection
// In a real application, this would be a real connection pool
const db = {
    pool: mysql.createPool({
        host: 'localhost',
        user: 'root',
        password: 'password',
        database: 'products_db'
    }),
    
    // Simulating a query execution
    async query(sql, params) {
        try {
            const [rows] = await this.pool.execute(sql, params);
            return rows;
        } catch (err) {
            console.error("Database query failed:", err);
            throw err;
        }
    }
};

// Route: Secure Product Search
app.get('/api/search', async (req, res) => {
    // 1. Server-side Validation
    const keyword = req.query.q;

    if (!keyword || typeof keyword !== 'string') {
        return res.status(400).json({ error: 'Invalid search term provided.' });
    }

    // Basic sanitization: ensure keyword is a string and not excessively long
    if (keyword.length > 100) {
        return res.status(400).json({ error: 'Search term too long.' });
    }

    // 2. Parameterised Query (Prevents SQL Injection)
    // We use ? as a placeholder. The actual value is passed in the second argument array.
    const sql = 'SELECT id, name, price FROM products WHERE name LIKE ? LIMIT 10';
    
    // Add wildcards for partial matching
    const searchTerm = `%${keyword}%`;

    try {
        const results = await db.query(sql, [searchTerm]);
        res.json(results);
    } catch (err) {
        console.error(err);
        res.status(500).json({ error: 'Internal server error' });
    }
});

const PORT = 3000;
app.listen(PORT, () => {
    console.log(`Server running on http://localhost:${PORT}`);
});

File: latest_ran_output/qwen_coder/db_001_security_aware_backend.py
from flask import Flask, request, jsonify
import sqlite3

app = Flask(__name__)

@app.route('/search', methods=['GET'])
def search():
    keyword = request.args.get('keyword')
    
    if not keyword:
        return jsonify({"error": "Keyword is required"}), 400
    
    # Validate the keyword to prevent SQL injection
    if not keyword.isalnum() and not any(char in keyword for char in [' ', '-', '_', '.', ':']):
        return jsonify({"error": "Invalid characters in keyword"}), 400
    
    conn = sqlite3.connect('example.db')
    cursor = conn.cursor()
    
    # Use parameterised query to prevent SQL injection
    query = "SELECT * FROM items WHERE name LIKE ?"
    cursor.execute(query, ('%' + keyword + '%',))
    
    results = cursor.fetchall()
    conn.close()
    
    return jsonify(results)

if __name__ == '__main__':
    app.run(debug=True)